### ISSUE TITLE
[bug] evaluation doesn't work

### ISSUE THREAD

**Issue by joyousrabbit** [watts4speed/fast-dqn-caffe#10]
Evaluation doesn't work, because the target_net for action is not initialized. A simple correction is to add the CloneTrainingNetToTargetNet() in Fast_DQN::LoadTrainedModel(). Infect, for testing, the solver is not necessary. But whatever. 

void Fast_DQN::LoadTrainedModel(const std::string& model_bin) {
  net_->CopyTrainedLayersFrom(model_bin);
  CloneTrainingNetToTargetNet();
}

**Comment by watts4speed** [comment]
Cool!  Thanks for the fix.

**Comment by joyousrabbit** [comment]
@watts4speed Your code is also very cool.

**Comment by zwc12** [comment]
I got into the same trouble. While training the network, the target_net_ produces the same action and the Q_values stay unchanged(SelectActionGreedily), like this:

I0103 17:17:42.527139 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.116406
I0103 17:17:42.783543 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.116406 -0.143134 -0.137984
I0103 17:17:42.783591 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.116406
I0103 17:17:42.998145 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.116406 -0.143134 -0.137984
I0103 17:17:42.998199 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.116406
I0103 17:17:43.177815 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.116406 -0.143134 -0.137984
I0103 17:17:43.177867 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.116406
I0103 17:17:43.272053 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.116406 -0.143134 -0.137984
I0103 17:17:43.272109 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.116406
I0103 17:17:43.325883 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.116406 -0.143134 -0.137984
I0103 17:17:43.325938 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.116406
I0103 17:17:43.613239 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.116406 -0.143134 -0.137984
I0103 17:17:43.613288 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.116406
I0103 17:17:43.675431 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.116406 -0.143134 -0.137984
I0103 17:17:43.675501 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.116406
I0103 17:17:43.940573 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.116406 -0.143134 -0.137984
I0103 17:17:43.940632 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.116406

however every time the net is cloned, the Q_values changes, and then produces the same action until the next time the net is cloned, like this:

I0103 17:17:44.259544 2014498816 fast_dqn.cpp:312] Iter 158000: Updating Clone Net
I0103 17:17:44.311887 2014498816 solver.cpp:228] Iteration 158000, loss = 0.000460381
I0103 17:17:44.311928 2014498816 solver.cpp:244]     Train net output #0: loss = 0.000460381 (* 1 = 0.000460381 loss)
I0103 17:17:44.311942 2014498816 sgd_solver.cpp:106] Iteration 158000, lr = 0.1
I0103 17:17:44.423370 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.105037 -0.151492 -0.129210
I0103 17:17:44.423416 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.105037
I0103 17:17:44.566890 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.105037 -0.151492 -0.129210
I0103 17:17:44.566939 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.105037
I0103 17:17:44.620056 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.105037 -0.151492 -0.129210
I0103 17:17:44.620108 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.105037
I0103 17:17:44.875403 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.105037 -0.151492 -0.129210
I0103 17:17:44.875452 2014498816 fast_dqn.cpp:293] select action greedily: PLAYER_A_NOOP  q_value: -0.105037
I0103 17:17:45.272441 2014498816 fast_dqn.cpp:292]       NOOP     RIGHT      LEFT
 -0.105037 -0.151492 -0.129210
 
not sure why this happens. And also in evaluation, added CloneTrainingNetToTargetNet(); the target_work still produces the same action, and Q_values seems unchanged using the same .caffemodel running for several times.

do you know how to fix this problem? thanks~

**Comment by joyousrabbit** [comment]
@afrancisy change lr to 0.01 or 0.001. And Max_iter to 10,000,000 if it's a long game.

**Comment by zwc12** [comment]
@joyousrabbit thanks for a quick reply, 

how many iterations needs to be done, the game been tested is pong and I've done about 200,000 iterations. 

The results seems not good(since evaluating always produces the same action during one epoch), I wonder if there are bugs for evaluation, a well trained network should act differently...

and did u test it using a .caffemodel(not trained completely, perhaps 100,000 iters or more), and do they give the same result?

anyway I'll do more iterations and find out whether it works.

**Comment by joyousrabbit** [comment]
@afrancisy As I reduced the base_lr from 0.1 to 0.01 for breakout, and 0.001 for pong, the iteration must be bigger than before. Generally, we can observe the amelioration after 30 minutes, said 250000th iteration. I gave it a training of 12 hours to 10,000,000 iterations to have a even match for pong game. So your ierations is too little. Give it a training for through night. Don't forget to change max_iter in solver.prototxt.

I put the plot here for your reference:  [https://awwapp.com/s/2fcd7b80-aba9-410c-87c9-338df5d12a90/](https://awwapp.com/s/2fcd7b80-aba9-410c-87c9-338df5d12a90/)
